osidb/sync_manager.py

```diff
@@ -0,0 +1,614 @@
+            self.last_consecutive_reschedules = 0
+            if (
+                self.last_consecutive_failures >= self.MAX_CONSECUTIVE_FAILURES
+                or permanent
+            ):
+                self.permanently_failed = True
+            self.save()
+        logger.info(f"{self.__class__.__name__} {self.sync_id}: Sync failed")
+        if self.permanently_failed:
+            logger.info(f"{self.__class__.__name__} {self.sync_id}: Sync permanently")
+        raise exception
+
+    def revoke_sync_task(self, celery_task):
+        """
+        Revoke sync_task.
+
+        :param celery_task: Associated Celery task.
+        """
+        self.update_synced_links()
+
+        celery_task.send_event("task-revoked")
+        logger.info(f"{self.__class__.__name__} {self.sync_id}: Revoked")
+        raise Ignore
+
+    def reschedule(self, reason):
+        """
+        Schedule sync_task to Celery queue again for a reason.
+
+        :param reason: Description for a reason why the sync_task was re-scheduled.
+        """
+        self.schedule()
+        with transaction.atomic():
+            self.refresh_from_db()
+            self.last_rescheduled_dt = timezone.now()
+            self.last_rescheduled_reason = reason
+            self.last_consecutive_reschedules += 1
+            self.save()
+        logger.info(
+            f"{self.__class__.__name__} {self.sync_id}: Sync re-scheduled ({reason})"
+        )
+
+    @classmethod
+    def get_sync_manager(cls, sync_id):
+        """
+        Returns sync manager object for a specific synchronized object ID.
+
+        :param sync_id: Unique ID for synchronized data object.
+        """
+        result, _ = cls.objects.get_or_create(sync_id=sync_id)
+        return result
+
+    @classmethod
+    def check_for_reschedules(cls):
+        """
+        This method needs to be called occasionally to check if any of the existing sync managers
+        need to re-schedule tasks for any reason (like previous failure).
+        """
+        for sync_manager in cls.objects.all():
+
+            # SCHEDULED, DID NOT START
+            # 1) Scheduled at least once before
+            # 2) Scheduled for more than MAX_SCHEDULE_DELAY
+            # 3) Not started after scheduled (or ever)
+            #
+            #      |       MAX_SCHEDULE_DELAY      |
+            #      |-------------------------------|---//-------?
+            #  Scheduled                          NOW        Started
+            #
+            if (
+                sync_manager.last_scheduled_dt is not None
+                and timezone.now() - sync_manager.last_scheduled_dt
+                > cls.MAX_SCHEDULE_DELAY
+                and (
+                    sync_manager.last_started_dt is None
+                    or sync_manager.last_started_dt < sync_manager.last_scheduled_dt
+                )
+            ):
+                sync_manager.reschedule("Sync did not start after MAX_SCHEDULE_DELAY")
```

Comment:
Is it intentional that the string is not an f-string and the constant value is not printed ?

Reply:
Yes.